src/main/java/dan200/computercraft/client/render/TileEntityMonitorRenderer.java

```diff
@@ -158,4 +141,92 @@ public void render( @Nonnull TileMonitor monitor, float partialTicks, @Nonnull M
 
         transform.pop();
     }
+
+    private static void renderTerminal( Matrix4f matrix, ClientMonitor monitor, float xMargin, float yMargin )
+    {
+        Terminal terminal = monitor.getTerminal();
+
+        MonitorRenderer renderType = MonitorRenderer.current();
+        boolean redraw = monitor.pollTerminalChanged();
+        if( monitor.createBuffer( renderType ) ) redraw = true;
+
+        switch( renderType )
+        {
+            case VBO:
+            {
+                VertexBuffer vbo = monitor.buffer;
+                if( redraw )
+                {
+                    Tessellator tessellator = Tessellator.getInstance();
+                    BufferBuilder builder = tessellator.getBuffer();
+                    builder.begin( FixedWidthFontRenderer.TYPE.getDrawMode(), FixedWidthFontRenderer.TYPE.getVertexFormat() );
+                    FixedWidthFontRenderer.drawTerminalWithoutCursor(
+                        IDENTITY, builder, 0, 0,
+                        terminal, !monitor.isColour(), yMargin, yMargin, xMargin, xMargin
+                    );
+
+                    builder.finishDrawing();
+                    vbo.upload( builder );
+                }
+
+                vbo.bindBuffer();
+                FixedWidthFontRenderer.TYPE.getVertexFormat().setupBufferState( 0L );
+                vbo.draw( matrix, FixedWidthFontRenderer.TYPE.getDrawMode() );
+                VertexBuffer.unbindBuffer();
+                FixedWidthFontRenderer.TYPE.getVertexFormat().clearBufferState();
+                break;
+            }
+
+            case TBO:
+            {
+                // TODO: Test whether we need this, or if 0 is enough. I'm curious as to how this plays with Optifine's
+                //  shaders.
+                int program = GlStateManager.getInteger( GL20.GL_CURRENT_PROGRAM );
```

Comment:
As a general rule, `glGet*` has crap performance, so I hope we don't need this in something as frequently invoked as `renderTerminal`. The whole point of `GlStateManager` is to cache the value of OpenGL state variables on the CPU to avoid these performance issues, does it not have a method for this?

Or does `GlStateManager.getInteger` already do that? IIRC it does not, but I'd be happily proven wrong.

Reply:
Hah, you wait 'til you see what we do for projection matrices! No `GlStateManager` doesn't, or at least not one I could find. Shaders aren't something MC uses much, and I think even that's been stripped out in 1.16.

Like the comment says, we may not need it. If this breaks with Optifine (which it most likely will), we can just not bother.

Reply:
Update: It breaks with Optifine, so we're not going to bother.